src/NUnitFramework/framework/Compatibility/System.Threading/SpinWait.cs

```diff
@@ -1,139 +1,363 @@
-                Thread.Sleep(ntime % step == 0 ? 1 : 0);
-#else
-                Thread.Yield ();
-#endif
-			}
-			else {
-				if (ntime % step == 0)
-#if NET20 || NET35
+        /// <summary>
+        /// Gets whether the next call to <see cref="SpinOnce"/> will yield the processor, triggering a
+        /// forced context switch.
+        /// </summary>
+        /// <value>Whether the next call to <see cref="SpinOnce"/> will yield the processor, triggering a
+        /// forced context switch.</value>
+        /// <remarks>
+        /// On a single-CPU machine, <see cref="SpinOnce"/> always yields the processor. On machines with
+        /// multiple CPUs, <see cref="SpinOnce"/> may yield after an unspecified number of calls.
+        /// </remarks>
+        public bool NextSpinWillYield
+        {
+            get { return m_count > YIELD_THRESHOLD || PlatformHelper.IsSingleProcessor; }
+        }
+
+        /// <summary>
+        /// Performs a single spin.
+        /// </summary>
+        /// <remarks>
+        /// This is typically called in a loop, and may change in behavior based on the number of times a
+        /// <see cref="SpinOnce"/> has been called thus far on this instance.
+        /// </remarks>
+        public void SpinOnce()
+        {
+            if (NextSpinWillYield)
+            {
+                //
+                // We must yield.
+                //
+                // We prefer to call Thread.Yield first, triggering a SwitchToThread. This
+                // unfortunately doesn't consider all runnable threads on all OS SKUs. In
+                // some cases, it may only consult the runnable threads whose ideal processor
+                // is the one currently executing code. Thus we oc----ionally issue a call to
+                // Sleep(0), which considers all runnable threads at equal priority. Even this
+                // is insufficient since we may be spin waiting for lower priority threads to
+                // execute; we therefore must call Sleep(1) once in a while too, which considers
+                // all runnable threads, regardless of ideal processor and priority, but may
+                // remove the thread from the scheduler's queue for 10+ms, if the system is
+                // configured to use the (default) coarse-grained system timer.
+                //
+
+                int yieldsSoFar = (m_count >= YIELD_THRESHOLD ? m_count - YIELD_THRESHOLD : m_count);
+
+                if ((yieldsSoFar % SLEEP_1_EVERY_HOW_MANY_TIMES) == (SLEEP_1_EVERY_HOW_MANY_TIMES - 1))
+                {
                     Thread.Sleep(1);
-#else
-                    Thread.Yield ();
-#endif
-                else
-                    // Multi-CPU system might be hyper-threaded, let other thread run
-                    Thread.SpinWait (Math.Min (ntime, maxTime) << 1);
-			}
-		}
+                }
+                else // if ((yieldsSoFar % SLEEP_0_EVERY_HOW_MANY_TIMES) == (SLEEP_0_EVERY_HOW_MANY_TIMES - 1))
```

Comment:
This is the only change of substance. `Thread.Yield()` isn't available in .NET 2.0/3.5. The original code used `Platform.Yield()` for those platforms, but that code is not available in Mono or in the reference source. The comments above say that yielding is preferable, but it still calls `Sleep(0)` for correctness, so I just made that the default.